upf/upf_pfcp_server.c

```diff
@@ -1228,11 +1232,29 @@ static uword
       /* Free messages that must be freed after the loop */
       pool_foreach (msg, psm->msg_pool)
       {
-	if (!msg->is_valid_pool_item ||
-	    (!hash_get (psm->free_msgs_by_node, msg->node) &&
-	     !hash_get (psm->free_msgs_by_sidx, msg->session_index)))
+	if (!msg->is_valid_pool_item)
```

Comment:
AFAIK this should never happen, a assert for the debug build and clib_warning for the release images might be better.

Reply:
Thanks, will fix that

Reply:
Looking further at this problem, this is not really a bug b/c `psm->msg_pool` is used in such way that there may be invalid messages in it. `pfcp_msg_pool_put()` doesn't use `pool_put()`, but rather sets `is_valid_pool_item` to `0` and then adds the message to a separate freelist, which is later used to populate `msg_pool_cache` for faster allocation. So, going through all of the msg pool with `pool_foreach()` is indeed quite inefficient (as noted below, I will fix that in a followup) but also it can stumble upon "invalid" messages in the pool.